plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/MiniSelector.java

```diff
@@ -0,0 +1,156 @@
+    this(container, apiFactory, null);
+  }
+
+  public MiniSelector(Composite container, IGoogleApiFactory apiFactory, Credential credential) {
+    this.credential = credential;
+    this.projectRepository = new ProjectRepository(apiFactory);
+    create(container);
+  }
+
+  private void create(Composite parent) {
+    displayExecutor = DisplayExecutor.create(parent.getDisplay());
+    comboViewer = new ComboViewer(parent, SWT.READ_ONLY | SWT.DROP_DOWN);
+    comboViewer.setComparator(new ViewerComparator());
+    comboViewer.setLabelProvider(new LabelProvider() {
+      @Override
+      public String getText(Object element) {
+        if (element instanceof GcpProject) {
+          GcpProject project = (GcpProject) element;
+          return project.getName() + " (" + project.getId() + ")";
+        }
+        return super.getText(element);
+      }
+
+    });
+    projectsProvider = new ProjectsProvider(projectRepository);
+    comboViewer.setContentProvider(projectsProvider);
+    comboViewer.setInput(credential);
+  }
+
+  public Control getControl() {
+    return comboViewer.getControl();
+  }
+
+  /**
+   * Set whether the controls are enabled or not.
+   */
+  public void setEnabled(boolean enabled) {
+    comboViewer.getControl().setEnabled(enabled);
+  }
+
+  public Credential getCredential() {
+    return credential;
+  }
+
+  public void setCredential(Credential credential) {
+    // try to preserve the selection if possible
+    GcpProject selected = getProject();
+    this.credential = credential;
+    comboViewer.setInput(credential);
+    if (selected != null) {
+      setProject(selected.getId());
+    }
+  }
+
+  @Override
+  public IStructuredSelection getSelection() {
+    return comboViewer.getStructuredSelection();
+  }
+
+  @Override
+  public void setSelection(ISelection selection) {
+    comboViewer.setSelection(selection);
+  }
+
+  @Override
+  public void addSelectionChangedListener(ISelectionChangedListener listener) {
+    comboViewer.addPostSelectionChangedListener(listener);
+  }
+
+  @Override
+  public void removeSelectionChangedListener(ISelectionChangedListener listener) {
+    comboViewer.removePostSelectionChangedListener(listener);
+  }
+
+  /**
+   * Return the currently selected project or {@code null} if none selected.
+   */
+  public GcpProject getProject() {
+    IStructuredSelection selection = comboViewer.getStructuredSelection();
+    return selection.isEmpty() ? null : (GcpProject) selection.getFirstElement();
+  }
+
+  /**
+   * Set the currently selected project.
+   */
+  public void setProject(GcpProject project) {
+    setProject(project.getId());
+  }
+
+  /**
+   * Set the currently selected project by ID; return the project or {@code null} if no project by
+   * that ID.
+   */
+  public void setProject(final String projectId) {
+    projectsProvider.resolve(projectId, new ProjectsProvider.Callback<GcpProject>() {
```

Comment:
`resolve()` is designed to execute the callback in its `displayExecutor`, so maybe it is an duplicate effort to run it in the `displayExecutor` of `MiniSelector` here?

Reply:
Oops, that's a mistake: `ProjectProvider` should be dispatching on its active thread.